core/src/main/java/com/linecorp/armeria/server/auth/AuthorizerChain.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.auth;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * Chains multiple {@link Authorizer}s together into a single {@link Authorizer}.
+ * Utilizes {@link AuthorizerSelectionStrategy} to select corresponding {@link AuthFailureHandler}.
+ */
+final class AuthorizerChain<T> implements Authorizer<T> {
+
+    enum AuthorizerSelectionStrategy {
+        /**
+         * Select a handler provided by the first {@link Authorizer} in the chain.
+         */
+        FIRST,
+        /**
+         * Select a handler provided by the last {@link Authorizer} in the chain.
+         */
+        LAST,
+        /**
+         * Select a handler provided by the first {@link Authorizer} in the chain that has an associated
+         * handler actually defined (non-NULL).
+         */
+        FIRST_WITH_HANDLER,
+        /**
+         * Select a handler provided by the last {@link Authorizer} in the chain that has an associated
+         * handler actually defined (non-NULL).
+         */
+        LAST_WITH_HANDLER
+    }
+
+    private final List<? extends Authorizer<T>> authorizers;
+    private final AuthorizerSelectionStrategy selectionStrategy;
+    @Nullable
+    private AuthSuccessHandler successHandler;
+    @Nullable
+    private AuthFailureHandler failureHandler;
```

Comment:
Yes, this chain processing does not look nice at the second look - we WILL have a problem with multiple requests. But not sure how we could make `successHandler` and `failureHandler` local as those exposed by `Authorizer` interface.
The only solution I can see right now is to return `successHandler` and `failureHandler` alongside the authorization status.

Reply:
> Yes, this chain processing does not look nice at the second look - we WILL have a problem with multiple requests. But not sure how we could make `successHandler` and `failureHandler` local as those exposed by `Authorizer` interface.
> The only solution I can see right now is to return `successHandler` and `failureHandler` alongside the authorization status.

Please advise on the best/optimal/acceptable solution.

Reply:
What about the following solution?

```
public interface Authorizer<T> {

    CompletionStage<Boolean> authorize(ServiceRequestContext ctx, T data);

    default CompletionStage<AuthorizationStatus> authorizeAndSupplyHandlers(ServiceRequestContext ctx, T data) {
        authorize(ctx, data).thenApply(status -> new AuthorizationStatusImpl(status));
    }

    ...
}


public interface AuthorizationStatus {

    boolean status();

    @Nullable
    AuthSuccessHandler successHandler();

    @Nullable
    AuthFailureHandler failureHandler();
}


class AuthorizationStatusImpl implements AuthorizationStatus {

    private final boolean status;
    @Nullable
    private final AuthSuccessHandler successHandler;
    @Nullable
    private final AuthFailureHandler failureHandler;

    AuthorizationStatusImpl(boolean status, @Nullable AuthSuccessHandler successHandler, @Nullable AuthFailureHandler failureHandler) {
        this.status = status;
        this.successHandler = successHandler;
        this.failureHandler = failureHandler;
    }

    AuthorizationStatusImpl(boolean status) {
        this(status, null, null);
    }

    boolean status() {
        return status;
    }

    AuthSuccessHandler successHandler() {
        return successHandler;
    }

    AuthFailureHandler failureHandler() {
        return failureHandler;
    }
}
```

This will allow removing `successHandler()/failureHandler()` from the `Authorizer` interface while keeping backward compatibility with the original interface. `AuthorizerChain` and `OAuth2` will implement the new `authorizeAndSupplyHandlers` method.

Reply:
I've implemented this approach in the latest commit. It looks fine. Please review.

Reply:
That sounds good to me. Let me take a close look soon.